File: smallestSWMBO.cpp
class Solution {
public:
    vector<int> smallestSubarrays(vector<int>& nums) {
        int n = nums.size();
        vector<int> lastSeen(32, -1);   // Track last index each bit appears
        vector<int> result(n);

        for (int i = n - 1; i >= 0; --i) {
            // Update last seen index for each bit present in nums[i]
            for (int bit = 0; bit < 32; ++bit) {
                if ((nums[i] >> bit) & 1) {
                    lastSeen[bit] = i;
                }
            }

            // Determine how far we must extend the subarray to capture all active bits
            int maxIdx = i;
            for (int bit = 0; bit < 32; ++bit) {
                if (lastSeen[bit] != -1) {
                    maxIdx = max(maxIdx, lastSeen[bit]);
                }
            }

            result[i] = maxIdx - i + 1;
        }

        return result;
    }
};

File: smallestSWMBO.ts
function smallestSubarrays(nums: number[]): number[] {
    const n = nums.length;
    const lastSeen: number[] = new Array(32).fill(-1); // Track last seen index for each bit
    const result: number[] = new Array(n).fill(0);

    for (let i = n - 1; i >= 0; i--) {
        // Update the last seen index for each bit in nums[i]
        for (let bit = 0; bit < 32; bit++) {
            if ((nums[i] >> bit) & 1) {
                lastSeen[bit] = i;
            }
        }

        // Determine the furthest index needed to reach the max OR
        let maxIdx = i;
        for (let bit = 0; bit < 32; bit++) {
            if (lastSeen[bit] !== -1) {
                maxIdx = Math.max(maxIdx, lastSeen[bit]);
            }
        }

        result[i] = maxIdx - i + 1;
    }

    return result;
}
